File: app/src/main/java/ru/hse/control_system_v2/connection/ConnectionClass.java
package ru.hse.control_system_v2.connection;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observable;

abstract class ConnectionClass {
    boolean isNeedToConnect;
    String connectionName;

    public void setNeedToConnect(boolean needToConnect) {
        isNeedToConnect = needToConnect;
    }

    public boolean isNeedToConnect() {
        return isNeedToConnect;
    }

    public String getConnectionName() {
        return connectionName;
    }

    public void setConnectionName(String connectionName) {
        this.connectionName = connectionName;
    }

    abstract void sentData();

    abstract void closeConnection();

    abstract @NonNull Observable<Object> openConnection();


}

File: app/src/main/java/ru/hse/control_system_v2/connection/ConnectionClass.kt
package ru.hse.control_system_v2.connection

import kotlinx.coroutines.Job
import ru.hse.control_system_v2.connection.data.classes.ConnectionDeviceModel

//класс обёртка сокета для одного устройства
abstract class ConnectionClass<T>(val connectionDeviceModel: ConnectionDeviceModel) {
    enum class ConnectionState {
        ON_ERROR,
        CONNECTING,
        ALIVE,
        DISABLED
    }

    var connectionState = ConnectionState.CONNECTING

    protected var socket: T? = null

    protected var coroutine: Job? = null

    abstract suspend fun sentData(data: ByteArray)
    abstract suspend fun closeConnection()
    abstract suspend fun openConnection()
    abstract suspend fun read(buffer: ByteArray, bytesRead: Int): Int
}